In [63]:
# Numerical computing library
# Used for fast array operations and numerical calculations
import numpy as np

# Data manipulation and analysis library
# Used for loading datasets, cleaning data, and tabular operations
import pandas as pd

# PyTorch library
# Used for tensor operations, autograd, and building neural networks
import torch

# Utility to split dataset into training and testing sets
# Ensures fair model evaluation
from sklearn.model_selection import train_test_split

# Feature scaling utility
# Standardizes features to have mean = 0 and std = 1
from sklearn.preprocessing import StandardScaler

# Label encoding utility
# Converts categorical labels into numerical form
from sklearn.preprocessing import LabelEncoder

In [64]:
# Load the Breast Cancer dataset directly from a GitHub raw URL
df = pd.read_csv(
    'https://raw.githubusercontent.com/gscdit/Breast-Cancer-Detection/refs/heads/master/data.csv'
)

# Display the first 5 rows of the dataset
print(df.head())


         id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness

In [65]:
df.shape

# 569 → number of rows (data samples)
# 33 → number of columns (features + labels)

(569, 33)

In [66]:
# Drop unnecessary columns from the DataFrame
df.drop(
    columns=['id', 'Unnamed: 32'],  # Columns to remove
    inplace=True                    # Modify the DataFrame directly
)

In [67]:
print(df.head())

  diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         M        17.99         10.38          122.80     1001.0   
1         M        20.57         17.77          132.90     1326.0   
2         M        19.69         21.25          130.00     1203.0   
3         M        11.42         20.38           77.58      386.1   
4         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33 

### Train Test Split

In [68]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    df.iloc[:, 1:],   # Features: all columns except the first one
    df.iloc[:, 0],    # Target: first column (label)
    test_size=0.2     # 20% data for testing, 80% for training
)

#### What each part means

1. `df.iloc[:, 1:]`

    * Selects all rows

    * Selects columns from index 1 onward

    * These are your input features (X)

2. `df.iloc[:, 0]`

    * Selects the first column

    * This is your target/label (y) (e.g., diagnosis)

3. `test_size=0.2`

    * 20% of data → test set

    * 80% of data → training set

### Scaling

In [69]:
# Create a StandardScaler object
# This scaler will normalize features to:
# mean = 0 and standard deviation = 1
scaler = StandardScaler()

# Fit the scaler ONLY on training data
# - Learns mean and std from X_train
# - Transforms X_train using those statistics
X_train = scaler.fit_transform(X_train)

# Transform the test data using the SAME scaler
# - Uses training mean and std
# - Does NOT re-fit (prevents data leakage)
X_test = scaler.transform(X_test)


#### ⚖️ Feature Scaling with StandardScaler

Feature scaling ensures that all input features are on a **similar scale**,
which is critical for gradient-based models.

---

#### What Is Standardization?

StandardScaler applies:
```
x_scaled = (x − mean) / standard_deviation
```

**After scaling:**

* Mean ≈ 0
* Standard deviation ≈ 1

**Without scaling:**

* Large features dominate gradients
* Training becomes unstable
* Slower convergence

**With scaling:**

* Faster learning
* Stable gradients
* Better model performance

**Models That Require Scaling**

✔ Logistic Regression

✔ Neural Networks

✔ SVM

✔ K-Means

❌ Tree-based models (Decision Trees, Random Forests)

**Learn scaling rules from training data $\rightarrow$ apply everywhere**

In [70]:
# X_train contains the training input features
# It is the output after:
# 1. Dropping unnecessary columns
# 2. Splitting data into train/test
# 3. Scaling features using StandardScaler

X_train

array([[-0.02330827, -1.41765487, -0.0381396 , ..., -0.18039409,
        -1.19845425, -0.61922761],
       [-0.16184858, -0.13718557, -0.15830126, ...,  0.4414885 ,
         0.97277489,  0.1388682 ],
       [-0.06948837, -0.81643094, -0.11936386, ..., -1.21253805,
        -1.67970607, -0.85707212],
       ...,
       [ 0.11523205, -0.64432485,  0.12263417, ...,  0.42186599,
         0.22931278,  0.19202081],
       [ 1.59010913,  3.01120847,  1.53987178, ...,  0.7403544 ,
        -0.31429393, -0.45010567],
       [-0.40429413, -0.50434523, -0.44384218, ..., -0.22552586,
         0.57626176, -0.57788669]])

In [71]:
X_train.shape

(455, 30)

## 🔍 Why It Looks Different Now

Earlier, `X_train` was a **DataFrame**:

```python
X_train = df.iloc[:, 1:]
```

After this line:

```python
X_train = scaler.fit_transform(X_train)
```

It became:

```text
numpy.ndarray
```

This is expected behavior of `StandardScaler`.

---

## 🧠 What the Values Mean

Each value in `X_train` is now a **z-score**:

```text
x_scaled = (x − mean_train) / std_train
```

So:

* Values around `0` → average
* Positive values → above average
* Negative values → below average

This makes gradients behave nicely during training.

---
## ✅ What `X_train` Represents (Now)

```python
X_train
```

### At this point in the pipeline, `X_train` is:

* ✅ **A NumPy array** (not a Pandas DataFrame anymore)
* ✅ **Scaled / standardized**
* ✅ **Training features only**
* ❌ **No labels**
* ❌ **No column names**

---

## 🔢 Shape Interpretation

If you check:

```python
X_train.shape
```

You’ll see something like:

```text
(455, 30)
```

Meaning:

* **455 samples** (80% of 569)
* **30 features** per sample

This tells you:

> Your neural network input layer must accept **30 features**

---


In [72]:
# y_train contains the target labels for the training set
# It was created by:
# 1. Selecting the label column from the DataFrame
# 2. Splitting the data into training and testing sets
#
# y_train is used to compute the loss during training
y_train

528    B
526    B
442    B
117    M
308    B
      ..
203    M
256    M
340    B
219    M
454    B
Name: diagnosis, Length: 455, dtype: object

In [73]:
y_train.shape

# One label per training sample
# Matches the number of rows in X_train
# X_train → model → predictions
# y_train → loss function

# The loss compares:
# Predictions vs y_train
# Gradients are computed to update weights

(455,)

# 🎯 Training Labels (`y_train`)

`y_train` represents the **ground-truth outputs** for the training data.
These values tell the model what the *correct answer* is.

---

## 1️⃣ What `y_train` Is

- A **1D array-like structure** (Pandas Series)
- Contains **labels only**
- Used **only during training**
- Corresponds exactly to `X_train`

---
So:

* `df.iloc[:, 0]` → label column
* `y_train` → 80% of those labels

🧠 Mental Model

> `X_train` = questions

> `y_train` = answers

**The model learns by reducing the difference between the two.**

### Label Encoding

In [74]:
# Create a LabelEncoder object
# This converts categorical labels into numeric form
encoder = LabelEncoder()

# Fit the encoder on training labels and transform them
# Example: ['M', 'B'] → [1, 0]
# IMPORTANT: fit ONLY on training labels
y_train = encoder.fit_transform(y_train)

# Transform test labels using the SAME encoder
# Ensures consistent label mapping
y_test = encoder.transform(y_test)

In [75]:
y_train

array([0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,

# 🔤 Label Encoding of Target Variables

Machine learning models require **numeric labels**.
This step converts categorical class labels into numbers.

---

## 1️⃣ What Is Label Encoding?

Label encoding maps each unique class to an integer.

Example:
```text
Benign (B)     → 0
Malignant (M)  → 1
```

**Resulting Data Type:**

After encoding:

* `y_train` $\rightarrow$ NumPy array
* `y_test` $\rightarrow$ NumPy array
*  Values are integers ($0$ or $1$)

### Numpy arrays to PyTorch tensors

In [76]:
# Convert NumPy arrays to PyTorch tensors
# This is required before feeding data into a PyTorch model

# Training features → PyTorch tensor
X_train_tensor = torch.from_numpy(X_train)

# Testing features → PyTorch tensor
X_test_tensor = torch.from_numpy(X_test)

# Training labels → PyTorch tensor
y_train_tensor = torch.from_numpy(y_train)

# Testing labels → PyTorch tensor
y_test_tensor = torch.from_numpy(y_test)


##  Converting NumPy Arrays to PyTorch Tensors

PyTorch models operate on **tensors**, not NumPy arrays.
This step converts all processed data into PyTorch-compatible format.

---



In [77]:
# Check the shape of the training feature tensor
# Format: (number_of_samples, number_of_features)
print(X_train_tensor.shape)

# Check the shape of the training label tensor
# Format: (number_of_samples,)
print(y_train_tensor.shape)

torch.Size([455, 30])
torch.Size([455])


## Why These Shapes Must Match

Each row in `X_train_tensor` must correspond to exactly one label in `y_train_tensor`.

`X_train_tensor[i] ↔ y_train_tensor[i]`


If this alignment breaks:
* Loss computation fails
* Training becomes incorrect or crashes

### Defining the model

In [78]:
class MySimpleNN():

    def __init__(self, X):
        """
        Constructor for the neural network.
        Initializes weights and bias.

        X: input feature tensor (used only to infer input dimension)
        """

        # Initialize weights randomly based on the number of features (X.shape[1]).
        # Each feature gets one weight  (num_features, 1)
        # requires_grad=True tells PyTorch to track operations on this tensor
        # so it can calculate gradients later (for backpropagation).
        self.weights = torch.rand(
            X.shape[1],  # number of input features
            1,
            dtype=torch.float64,
            requires_grad=True
        )

        # Initialize bias to zero as a single scalar. This acts as the intercept.
        self.bias = torch.zeros(
            1,
            dtype=torch.float64,
            requires_grad=True
        )

    def forward(self, X):
        """
        Forward pass of the model.
        Computes predictions from input features.
        Math: y_pred = sigmoid(X * weights + bias)
        """

        # 1. Linear Step: Matrix multiplication (X @ weights) + bias
        # z represents the "logits" or raw scores before activation.
        z = torch.matmul(X, self.weights) + self.bias

        # 2. Activation Step: Apply the Sigmoid function.
        # This squashes the output between 0 and 1 (interpretable as probability).
        # Apply sigmoid activation to convert logits to probabilities
        y_pred = torch.sigmoid(z)

        return y_pred

    def loss_function(self, y_pred, y):
        """
        Calculate Binary Cross-Entropy loss function.
        This measures how well the predicted probabilities match the actual labels.
        y_pred: predicted probabilities
        y: true labels
        """

        # Small value to prevent log(0) which causes -inf
        # Clamp predictions to prevent 'inf' errors.
        # log(0) is undefined (-infinity), so we force values to be
        # slightly larger than 0 and slightly smaller than 1.
        epsilon = 1e-7

        # Clamp predictions to (epsilon, 1 - epsilon)
        y_pred = torch.clamp(y_pred, epsilon, 1 - epsilon)

        # Calculate the Negative Log Likelihood (BCE Formula):
        # Loss = -1/N * sum( y*log(y_pred) + (1-y)*log(1-y_pred) )
        loss = -(
            y * torch.log(y_pred) +
            (1 - y) * torch.log(1 - y_pred)
        ).mean()

        return loss


### 2. Detailed Breakdown

#### 🧠 A Simple Neural Network (From Scratch)

This class implements a **single-layer neural network** for binary classification.
Mathematically, this is equivalent to **logistic regression**.

---

## 1️⃣ Model Structure

The model consists of:
- A linear layer
- A sigmoid activation
- A binary cross-entropy loss

```text
X → (XW + b) → sigmoid → ŷ → BCE loss
```

#### A. `__init__`: Setting the Stage

This method initializes the "learnable" parts of your network.

* **`self.weights`**: You are initializing random weights. If your input `X` has 10 features (columns), you need 10 weights.
* **`requires_grad=True`**: This is the magic switch in PyTorch. It turns on **Autograd**. It tells PyTorch, *"Watch these variables. If I use them in a calculation, remember the steps so you can calculate the gradient (slope) later."* This is essential for training.

#### B. `forward`: Making a Prediction

This defines how data flows through the network.

* **`torch.matmul(X, self.weights)`**: This is the dot product. It multiplies every input feature by its corresponding weight and sums them up.
* **`torch.sigmoid(z)`**: A neural network without an activation function is just Linear Regression. The Sigmoid function is non-linear; it forces the output to strictly stay between  and .
* If  is a large positive number, Sigmoid .
* If  is a large negative number, Sigmoid .



#### C. `loss_function`: Grading the Performance

You are manually implementing **Binary Cross Entropy (BCE)**, which is the standard cost function for binary classification.

* **The Clamping (`epsilon`)**: This is a numerical stability trick.
* If the model predicts exactly `0` but the label is `1`, the math asks for `log(0)`, which is negative infinity. This causes the program to crash or return `NaN`.
* `clamp` ensures the value is never exactly 0, but rather `0.0000001`.


* **The Formula**:


* If the actual label  is 1, only the first part of the equation matters (). We want  to be close to 1 to minimize loss.
* If the actual label  is 0, only the second part matters ().




### Important Parameters

In [79]:
# Learning rate controls how big each parameter update step is
# Too large → unstable training
# Too small → very slow learning
learning_rate = 0.1

# Number of times the model will see the entire training dataset
# One epoch = one full forward + backward pass over the training data
epochs = 25

# Learning rate = how big each step is
# Epochs = how many steps we take

### Training Pipeline

In [80]:
# -------------------------------
# Create the model
# -------------------------------

# Initialize the neural network
# X_train_tensor is passed only to determine input feature size
model = MySimpleNN(X_train_tensor)


# -------------------------------
# Training loop
# -------------------------------

for epoch in range(epochs):

    # -------- Forward pass --------
    # Compute model predictions for training data
    y_pred = model.forward(X_train_tensor)

    # -------- Loss computation --------
    # Compare predictions with true labels
    loss = model.loss_function(y_pred, y_train_tensor)

    # -------- Backward pass --------
    # Compute gradients of loss w.r.t. weights and bias
    loss.backward()

    # -------- Parameter update (Gradient Descent) --------
    # Disable gradient tracking while updating parameters
    with torch.no_grad():
        model.weights -= learning_rate * model.weights.grad
        model.bias -= learning_rate * model.bias.grad

    # -------- Zero gradients --------
    # Gradients accumulate by default, so we must reset them
    model.weights.grad.zero_()
    model.bias.grad.zero_()

    # -------- Logging --------
    # Print loss for monitoring training progress
    print(f'Epoch: {epoch + 1}, Loss: {loss.item()}')


Epoch: 1, Loss: 3.6075409078505487
Epoch: 2, Loss: 3.479149891634001
Epoch: 3, Loss: 3.339358523495705
Epoch: 4, Loss: 3.198969996853641
Epoch: 5, Loss: 3.0556455601320143
Epoch: 6, Loss: 2.907397226325239
Epoch: 7, Loss: 2.743784816742959
Epoch: 8, Loss: 2.5797988405603918
Epoch: 9, Loss: 2.4136810626776803
Epoch: 10, Loss: 2.247786207276684
Epoch: 11, Loss: 2.080582097019753
Epoch: 12, Loss: 1.9129719899687871
Epoch: 13, Loss: 1.7497833316498423
Epoch: 14, Loss: 1.5967493688579615
Epoch: 15, Loss: 1.4552639756889039
Epoch: 16, Loss: 1.3225403526205357
Epoch: 17, Loss: 1.205321156552129
Epoch: 18, Loss: 1.10483123180394
Epoch: 19, Loss: 1.021677641881896
Epoch: 20, Loss: 0.9555485719621205
Epoch: 21, Loss: 0.9050139750249961
Epoch: 22, Loss: 0.8676429056209236
Epoch: 23, Loss: 0.8404881640640138
Epoch: 24, Loss: 0.8206826196522852
Epoch: 25, Loss: 0.8058568434316598


# 🔁 Training Loop: Manual Gradient Descent with Autograd

This training loop implements **end-to-end learning** for a simple neural network
using PyTorch’s automatic differentiation.

---

## 1️⃣ Model Creation

```python
model = MySimpleNN(X_train_tensor)
````

* Initializes weights and bias
* Input size is inferred from `X_train_tensor.shape[1]`
* Parameters have `requires_grad=True`

---

## 2️⃣ Epoch Loop

```python
for epoch in range(epochs):
```

An **epoch** is one complete pass through the training dataset.

---

## 3️⃣ Forward Pass

```python
y_pred = model.forward(X_train_tensor)
```

Computes predictions using:

```text
z = XW + b
ŷ = sigmoid(z)
```

---

## 4️⃣ Loss Computation

```python
loss = model.loss_function(y_pred, y_train_tensor)
```

* Uses Binary Cross-Entropy (BCE)
* Measures how wrong predictions are
* Returns a **scalar loss** (required for `.backward()`)

---

## 5️⃣ Backward Pass

```python
loss.backward()
```

Autograd:

* Traverses the computation graph backward
* Applies the chain rule
* Computes gradients:

  * `model.weights.grad`
  * `model.bias.grad`

---

## 6️⃣ Parameter Update (Gradient Descent)

```python
with torch.no_grad():
    model.weights -= learning_rate * model.weights.grad
    model.bias -= learning_rate * model.bias.grad
```

### Gradient Descent Rule

```text
θ = θ − α · ∇θ
```

Where:

* `θ` = parameter
* `α` = learning rate
* `∇θ` = gradient

`torch.no_grad()` prevents autograd from tracking updates.

---

## 7️⃣ Zeroing Gradients (CRITICAL)

```python
model.weights.grad.zero_()
model.bias.grad.zero_()
```

Why?

* Gradients **accumulate by default**
* Not clearing them causes incorrect updates

Equivalent to:

```python
optimizer.zero_grad()
```

---

## 8️⃣ Monitoring Training

```python
print(f"Epoch: {epoch + 1}, Loss: {loss.item()}")
```

* `loss.item()` converts tensor → Python float
* Loss should **decrease over epochs**

---

## 🧠 Mental Model

> **Forward → Loss → Backward → Update → Repeat**

This loop is the heart of *all* neural network training.

---

## 9️⃣ How This Maps to Real PyTorch Code

Manual version:

```python
loss.backward()
weights -= lr * weights.grad
weights.grad.zero_()
```

PyTorch version:

```python
optimizer.zero_grad()
loss.backward()
optimizer.step()
```

---

## ✅ One-Line Summary

> This loop trains a neural network by repeatedly computing loss, gradients, and updating parameters using gradient descent.



In [81]:
# Access the bias parameter of the model
# This is a learnable scalar that shifts the decision boundary
model.bias

# Weights decide direction
# Bias decides starting point


tensor([-0.1514], dtype=torch.float64, requires_grad=True)

# 🎯 Model Bias (`model.bias`)

`model.bias` is a **learnable parameter** of the neural network that controls
the **offset** of the decision boundary.

---

## 1️⃣ What Is the Bias?

In the linear equation:

```text
z = XW + b
```
* W → weights (feature importance)
* b → bias (baseline shift)

The bias allows the model to make predictions even when:
x = 0

### Evaluation

In [82]:
# Model evaluation (no training)

# Disable gradient tracking to save memory and speed up computation
with torch.no_grad():

    # Forward pass on the test data
    # Produces predicted probabilities in range (0, 1)
    y_pred = model.forward(X_test_tensor)

    # Convert probabilities to class labels
    # Threshold = 0.9
    # Values > 0.9 → 1, else → 0
    y_pred = (y_pred > 0.9).float()

    # Compare predictions with true labels
    # (True → 1.0, False → 0.0)
    accuracy = (y_pred == y_test_tensor).float().mean()

    # Print accuracy as a Python float
    print(f'Accuracy: {accuracy.item()}')


Accuracy: 0.6141889691352844
